src/python/pants/backend/go/util_rules/binary.py

```diff
@@ -62,15 +66,24 @@ async def determine_main_pkg_for_go_binary(
             WrappedTarget,
             WrappedTargetRequest(specified_address, description_of_origin=description_of_origin),
         )
-        if not wrapped_specified_tgt.target.has_field(GoPackageSourcesField):
+        if not wrapped_specified_tgt.target.has_field(
+            GoPackageSourcesField
+        ) and not wrapped_specified_tgt.target.has_field(GoImportPathField):
```

Comment:
I think this can check that it only has `GoImportPathField`, right? That's more obvious. And then you'll still have the `GoPackageSourcesField` check below.

Reply:
Technically, but this catches the case where it actually isn't a go package at all. If we only check for ImportPath anything that doesn't have sources will be assumed to be an external go package. I'll see if I can repro that case again.

Reply:
Ah no, it's the other way around. We have to check if it has none of the fields. You can specify a first-party go_package without `import_path`.

Reply:
> You can specify a first-party go_package without import_path.

Common misunderstanding with the target API. `tgt.has_field()` checks if the field is _registered_ on that target type, regardless of whether a user explicitly set the field or not in a BUILD file.

So a first party `go_package` _always_ has `GoImportPathField`